In [2]:
import matplotlib.pyplot as plt

In [3]:
!pip install tensorflow==2.7.0

In [4]:
!pip install --upgrade jedi==0.17.2

In [5]:
import pandas as pd
import seaborn as sns


Implementation 

The dataset in the table 4.1  represents structured dataset of type-2 diabetic laboratory test from the  Indian National Institute of Diabetes and Digestive and Kidney Diseases, for the year 2018/2019.  In this study, an exploratory data analysis was performed to better understand the varibles in the dataset, clean the data from noise/redundancies, and to identify the relatiOnship between the variables. From the dataset, the varibles required to assess the outcomes of diabetic tests includes;  to determine the Pregnancy status and duration, the 	Glucose level,Blood Pressure, Patients	Skin Thickness,	Insulin level, body mass index	(BMI), the	Diabetes Pedigree Function, and the Age of patient. 
 

In [6]:
diabetes_lab_dataset = pd.read_csv("C:\\Users\\USER\\Desktop\\diabetes.csv")
diabetes_lab_dataset

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
746            1      147             94             41        0  49.3   
747            1       81             74             41       57  46.3   
748            3      187             70             22      200  36.4   
749            6      162             62              0        0  24.3   
750            4      136             70              0        0  31.2   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
746                     0.358   27        1  
747                     1.096   32        0  
748                     0.408   36        1  
749                     0.178   50        1  
750                     1.182   22        1  

[751 rows x 9 columns]

In [7]:
diabetes_test_set =pd.read_csv("C:\\Users\\USER\\Desktop\\DiabetesTest.csv")
diabetes_test_set


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             1      121             78             39       74  39.0   
1             3      108             62             24        0  26.0   
2             0      181             88             44      510  43.3   
3             8      154             78             32        0  32.4   
4             1      128             88             39      110  36.5   
5             7      137             90             41        0  32.0   
6             0      123             72              0        0  36.3   
7             1      106             76              0        0  37.5   
8             6      190             92              0        0  35.5   
9             2       88             58             26       16  28.4   
10            9      170             74             31        0  44.0   
11            9       89             62              0        0  22.5   
12           10      101             76             48      180  32.9   
13            2      122             70             27        0  36.8   
14            5      121             72             23      112  26.2   
15            1      126             60              0        0  30.1   
16            1       93             70             31        0  30.4   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.261   28        0  
1                      0.223   25        0  
2                      0.222   26        1  
3                      0.443   45        1  
4                      1.057   37        1  
5                      0.391   39        0  
6                      0.258   52        1  
7                      0.197   26        0  
8                      0.278   66        1  
9                      0.766   22        0  
10                     0.403   43        1  
11                     0.142   33        0  
12                     0.171   63        0  
13                     0.340   27        0  
14                     0.245   30        0  
15                     0.349   47        1  
16                     0.315   23        0

In [8]:
diabetes_lab_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   751.000000  751.000000     751.000000     751.000000  751.000000   
mean      3.844208  120.757656      68.984021      20.462051   80.271638   
std       3.367842   32.003164      19.495839      15.925772  114.979909   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   36.000000   
75%       6.000000  140.500000      80.000000      32.000000  129.500000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  751.000000                751.000000  751.000000  751.000000  
mean    31.958056                  0.474487   33.151798    0.347537  
std      7.922160                  0.332903   11.708156    0.476505  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.244000   24.000000    0.000000  
50%     32.000000                  0.378000   29.000000    0.000000  
75%     36.550000                  0.629500   40.500000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

Table 4.1 has 768 rows, 9 columns. interstingly, the variables has zero null values, i.e., all the features entries were adequately represented. However, the aim of this study is to discover the features patterns which will accurately depict a diabetic status at reduced run time, it is important to understand the statistical distribution of each variables in dataset. 

In [9]:
diabetes_lab_dataset.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

The outcome label is a nominal variable as it contains binary values [1,0] to determine the  diabetic status of an individual. The data contains 268 positive diabetes laboratory tests, while 500 tests indicates negative results . What is yet to be established is the interval between binary status, i.e. the possibility of transforming the outcome from nominal to a continous variable. The exploratory analysis will be limited to continous and discrete valued varibles, as these variables enable high number of unique values that is expedient to improve the accuracy models  accuracies.   The diabetes pedigree function variable which describes the family genetic history of an individual to diabetes, has the highest (517) distinct number of observations. 

In [10]:

diabetes_lab_dataset_mean = diabetes_lab_dataset.mean()
diabetes_lab_dataset_std = diabetes_lab_dataset.std()
diabetes_test_set_mean = diabetes_test_set.mean()
diabetes_test_set_std = diabetes_test_set.std()
diabetes_training_dataset_norm = (diabetes_lab_dataset-diabetes_lab_dataset_mean)/ diabetes_lab_dataset_std
diabetes_test_set_norm = (diabetes_test_set- diabetes_test_set_mean)/ diabetes_test_set_std
#roc_train = diabetes_lab_dataset['Outcome'].copy()
diabetes_training_dataset_norm['Outcome'] = diabetes_lab_dataset['Outcome'].astype(float)
#diabetes_training_dataset_norm['Real_Outcome']
#roc_test = diabetes_test_set['Outcome'].copy()
diabetes_test_set_norm['Outcome'] = diabetes_test_set['Outcome'].astype(float)
#diabetes_test_set_norm['Real_Outcome']
#diabetes_test_set_norm.drop('Outcome', axis=1, inplace=True)
diabetes_test_set_norm
#diabetes_training_dataset_norm.drop('Outcome', axis=1, inplace=True)
diabetes_training_dataset_norm

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.640111  0.851239       0.154699       0.912857 -0.698136  0.207260   
1      -0.844519 -1.117316      -0.153059       0.536109 -0.698136 -0.676338   
2       1.233963  1.944881      -0.255645      -1.284839 -0.698136 -1.092891   
3      -0.844519 -0.992329      -0.153059       0.159361  0.119398 -0.486995   
4      -1.141445  0.507523      -1.486677       0.912857  0.762989  1.406427   
..           ...       ...            ...            ...       ...       ...   
746    -0.844519  0.819992       1.283144       1.289605 -0.698136  2.189042   
747    -0.844519 -1.242304       0.257285       1.289605 -0.202397  1.810358   
748    -0.250667  2.069869       0.052113       0.096570  1.041298  0.560699   
749     0.640111  1.288696      -0.358231      -1.284839 -0.698136 -0.966663   
750     0.046259  0.476276       0.052113      -1.284839 -0.698136 -0.095688   

     DiabetesPedigreeFunction       Age  Outcome  
0                    0.458129  1.439014      1.0  
1                   -0.370941 -0.183786      0.0  
2                    0.593303 -0.098376      1.0  
3                   -0.923654 -1.037892      0.0  
4                    5.447566 -0.012965      1.0  
..                        ...       ...      ...  
746                 -0.349913 -0.525428      1.0  
747                  1.866947 -0.098376      0.0  
748                 -0.199720  0.243267      1.0  
749                 -0.890611  1.439014      1.0  
750                  2.125280 -0.952481      1.0  

[751 rows x 9 columns]

From the dataset, it is implied that patient with the diabetes pedigree function of  average mean value of 0.55 have the great risk of being diabetic, and also an individual with less than 0.088 pedigree function value is at reduced diabetes risk. The 50% percentile rank of pedigree function that results in positive diabetic status is between the value range 0.2625 - 0.72800.  The pedigree function values is a stable variable that can be used to dtermine  a diabetic status,but the real-valued attribute tends discrete,  as result of the fact that it will take much period of time for a family diabetic history to be altered. For instance,  it is convetionaly rare to identify an individual with a low valued pedigree function (<0.088) to test positive to diabetes. Hence, the diabetic  posistive status of patient  cannot be solely determined by the pedigree function, due to the fact a discplined individual can disrupt the trend. The next ranked high unique valued variables are; Body mass Index,Insulin, and Glucose.  (Bays et.al,2007) has argued that the increased BMI was associated with increased prevalence of diabetes, while the glucose variable  measures the  blood sugar level for diabetic testing. Considering the distribution of the glucose feature of average mean value of 141.257, with high standard deviation of 31.93, and the insulin parameter has average mean value of 79.8, with standard deviation values of 115.24. The average mean  values of both variables seem to be less  robust and unreliable to determining a accurate diabetic status. Conversely, in the relationship analysis of all the variables  the glucose variable the is the most correlated feature in respect to the diabetic outcome at the statistical significance of 0.47. 

Hence, this study adopts the BMI with the diabetic test outcome to perform the realtionship analysis in respect to other variables.


Refrence
Bays, H. E., Chapman, R. H., & Grandy, S. (2007). The relationship of body mass index to diabetes mellitus, hypertension and dyslipidaemia: comparison of data from two national surveys. International Journal of Clinical Practice, 61(5), 737–747. https://doi.org/10.1111/j.1742-1241.2007.01336.x

In [11]:
diabetes_training_dataset_norm

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.640111  0.851239       0.154699       0.912857 -0.698136  0.207260   
1      -0.844519 -1.117316      -0.153059       0.536109 -0.698136 -0.676338   
2       1.233963  1.944881      -0.255645      -1.284839 -0.698136 -1.092891   
3      -0.844519 -0.992329      -0.153059       0.159361  0.119398 -0.486995   
4      -1.141445  0.507523      -1.486677       0.912857  0.762989  1.406427   
..           ...       ...            ...            ...       ...       ...   
746    -0.844519  0.819992       1.283144       1.289605 -0.698136  2.189042   
747    -0.844519 -1.242304       0.257285       1.289605 -0.202397  1.810358   
748    -0.250667  2.069869       0.052113       0.096570  1.041298  0.560699   
749     0.640111  1.288696      -0.358231      -1.284839 -0.698136 -0.966663   
750     0.046259  0.476276       0.052113      -1.284839 -0.698136 -0.095688   

     DiabetesPedigreeFunction       Age  Outcome  
0                    0.458129  1.439014      1.0  
1                   -0.370941 -0.183786      0.0  
2                    0.593303 -0.098376      1.0  
3                   -0.923654 -1.037892      0.0  
4                    5.447566 -0.012965      1.0  
..                        ...       ...      ...  
746                 -0.349913 -0.525428      1.0  
747                  1.866947 -0.098376      0.0  
748                 -0.199720  0.243267      1.0  
749                 -0.890611  1.439014      1.0  
750                  2.125280 -0.952481      1.0  

[751 rows x 9 columns]

In [12]:
diabetes_lab_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   751.000000  751.000000     751.000000     751.000000  751.000000   
mean      3.844208  120.757656      68.984021      20.462051   80.271638   
std       3.367842   32.003164      19.495839      15.925772  114.979909   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   36.000000   
75%       6.000000  140.500000      80.000000      32.000000  129.500000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  751.000000                751.000000  751.000000  751.000000  
mean    31.958056                  0.474487   33.151798    0.347537  
std      7.922160                  0.332903   11.708156    0.476505  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.244000   24.000000    0.000000  
50%     32.000000                  0.378000   29.000000    0.000000  
75%     36.550000                  0.629500   40.500000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

  Moreover, the fact that the varibles can be statistically represented, does not mean they are all relevant to the predicitve evaluation of this study.  The non-required varibles can be serached through relationship analysis of the diabetic features, and hence will be droped.  The 

In [13]:
from sklearn.model_selection import train_test_split
# Split dataset
X= diabetes_lab_dataset.drop(columns=['Outcome'])
X = X.drop(columns=['SkinThickness'])
y = diabetes_lab_dataset['Outcome']
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)
X_train,X_val,y_train,y_val = train_test_split(X_train,y_train,test_size=0.2,random_state=1)

In [14]:
import tensorflow as tf
from tensorflow.keras import layers
from tensorflow.keras.utils import plot_model
from tensorflow.keras.metrics import Precision,AUC,Recall,Accuracy

feature_columns = []
diabetes_training_dataset_norm.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [15]:
Pregnancies = tf.feature_column.numeric_column("Pregnancies")
feature_columns.append(Pregnancies)
Glucose= tf.feature_column.numeric_column("Glucose")
feature_columns.append(Glucose)
BloodPressure = tf.feature_column.numeric_column("BloodPressure")
feature_columns.append(BloodPressure)
SkinThickness =tf.feature_column.numeric_column("SkinThickness")
feature_columns.append(SkinThickness)
Insulin = tf.feature_column.numeric_column("Insulin")
feature_columns.append(Insulin)
BMI = tf.feature_column.numeric_column("BMI")
feature_columns.append(BMI)
DiabetesPedigreeFunction = tf.feature_column.numeric_column("DiabetesPedigreeFunction")
feature_columns.append(DiabetesPedigreeFunction)
Age = tf.feature_column.numeric_column("Age")
feature_columns.append(Age)
Outcome = tf.feature_column.numeric_column("Outcome")
feature_columns.append(Outcome)

In [16]:

fp_layer = tf.keras.layers.DenseFeatures(feature_columns)


In [17]:
!pip install tensorboard

In [18]:
def create_model(lrn_rate,input_layer):
    model = tf.keras.models.Sequential()
    model.add(input_layer)
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden1'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden2'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden3'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden4'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden5'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden6'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden7'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden8'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden9'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden10'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden11'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden12'))
    model.add(tf.keras.layers.Dense(units=36,activation='relu',name='Hidden13'))
    model.add(tf.keras.layers.Dense(units=1,activation='softmax',name='ouput'))
    model.compile(optimizer = tf.keras.optimizers.RMSprop(learning_rate = lrn_rate),loss=tf.keras.losses.BinaryCrossentropy(), 
                  metrics =[Accuracy(name='acc')])
    return model

In [19]:
%reload_ext tensorboard

In [20]:
from tensorflow.keras.callbacks import TensorBoard

In [21]:
callbacks=TensorBoard(log_dir='logs\\sgd', histogram_freq=1)

In [22]:
# train model
import numpy as np

def train_model(dataset,lab_name,epoch,batch_size,model):
    features = {name:np.array(value) for name,value in dataset.items()}
    feat = features.copy()
    label = np.array(features.pop(lab_name))
    model.fit(x=feat, y=label,batch_size=batch_size,epochs=epoch,validation_split=0.35,shuffle=True,callbacks=callbacks)
    %tensorboard  --logdir='logs\\sgd' --port 5000
    
    return 
    

In [23]:
diabetes_training_dataset_norm = diabetes_training_dataset_norm.reindex(np.random.permutation(diabetes_training_dataset_norm.index))

In [24]:
lrate = 0.001
epoch = 170
batch_size = 350
lb_name = "Outcome"

mlp_diabetes_model = create_model(lrate,fp_layer)
train_model(diabetes_training_dataset_norm,lb_name,epoch,batch_size,mlp_diabetes_model)

feat_test = {name:np.array(value) for name,value in diabetes_test_set_norm.items() }

test_feat = feat_test.copy()
tester_label =np.array(test_feat.pop(lb_name)) 
mlp_diabetes_model.evaluate(x=feat_test,y=tester_label,batch_size=batch_size)

Epoch 1/170
Consider rewriting this model with the Functional API.
Consider rewriting this model with the Functional API.
1/2 [==============>...............] - ETA: 2s - loss: 0.6935 - acc: 0.3514WARNING:tensorflow:Layers in a Sequential model should only have a single input tensor, but we receive a <class 'dict'> input: {'Pregnancies': <tf.Tensor 'IteratorGetNext:7' shape=(None,) dtype=float32>, 'Glucose': <tf.Tensor 'IteratorGetNext:4' shape=(None,) dtype=float32>, 'BloodPressure': <tf.Tensor 'IteratorGetNext:2' shape=(None,) dtype=float32>, 'SkinThickness': <tf.Tensor 'IteratorGetNext:8' shape=(None,) dtype=float32>, 'Insulin': <tf.Tensor 'IteratorGetNext:5' shape=(None,) dtype=float32>, 'BMI': <tf.Tensor 'IteratorGetNext:1' shape=(None,) dtype=float32>, 'DiabetesPedigreeFunction': <tf.Tensor 'IteratorGetNext:3' shape=(None,) dtype=float32>, 'Age': <tf.Tensor 'IteratorGetNext:0' shape=(None,) dtype=float32>, 'Outcome': <tf.Tensor 'IteratorGetNext:6' shape=(None,) dtype=float32>}
Co

Reusing TensorBoard on port 5000 (pid 7916), started 12:35:21 ago. (Use '!kill 7916' to kill it.)

1/1 [==============================] - 0s 42ms/step - loss: 1.0231e-05 - acc: 0.4118


[1.0231310625385959e-05, 0.4117647111415863]

In [25]:
!pip install pydot

In [26]:
!pip install graphviz

In [27]:
from tensorflow.keras.utils import plot_model
plot_model(mlp_diabetes_model,to_file="mlp_diabetes_model.png",show_shapes=False)

('You must install pydot (`pip install pydot`) and install graphviz (see instructions at https://graphviz.gitlab.io/download/) ', 'for plot_model/model_to_dot to work.')


In [28]:
mlp_diabetes_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_features (DenseFeatur  multiple                 0         
 es)                                                             
                                                                 
 Hidden1 (Dense)             multiple                  360       
                                                                 
 Hidden2 (Dense)             multiple                  1332      
                                                                 
 Hidden3 (Dense)             multiple                  1332      
                                                                 
 Hidden4 (Dense)             multiple                  1332      
                                                                 
 Hidden5 (Dense)             multiple                  1332      
                                                        